### PR TITLE
Update Proof Creation Interfaces

### PR DESCRIPTION
**Pull request by iamyulong**
## Summary

This PR cleans up proof creation interfaces, based on the latest team decisions.  

Table below summarizes the changes.

### Scrypto

| Blueprint | Method          | Change |
|-----------|--------------------------|---------------------------------|
| Bucket    | `create_proof`           | Now creates a proof of amount `1`          |
|           | `create_proof_of_amount`    | New      |
|           | `create_proof_of_non_fungibles` | New|
|           | `create_proof_of_all`  | New |
| Vault     | `create_proof`           | Now creates a proof of amount `1`          |
|           | `create_proof_of_amount` | Renamed        |
|           | `create_proof_of_non_fungibles`    | Renamed  |
| AuthZone  | `create_proof`           | Now creates a proof of amount `1`          |
|           | `create_proof_of_amount` | Renamed        |
|           | `create_proof_of_non_fungibles`    | Renamed |
|           | `create_proof_of_all`    | New |
| Account  | `create_proof`           | Now creates a proof of amount `1`          |
|           | `create_proof_of_amount` | Renamed        |
|           | `create_proof_of_non_fungibles`    | Renamed |

Note that `authorize` will be changed to `authorize`, `authorize_with_amount`, `authorize_with_non_fungibles`, `authorize_with_all`, to be consistent.

### Manifest

| Method                         | Change          |
|-----------------------------------------|-----------------------------------------------------------|
| `CREATE_PROOF_FROM_BUCKET`              | Now creates a proof of amount `1`                                    |
| `CREATE_PROOF_FROM_BUCKET_OF_AMOUNT`              | New                                    |
| `CREATE_PROOF_FROM_BUCKET_OF_NON_FUNGIBLES`              | New |
| `CREATE_PROOF_FROM_BUCKET_OF_ALL`              | New |
| `CREATE_PROOF_FROM_AUTH_ZONE`           | Now creates a proof of amount `1`                                    |
| `CREATE_PROOF_FROM_AUTH_ZONE_OF_AMOUNT` | Renamed    |
| `CREATE_PROOF_FROM_AUTH_ZONE_OF_NON_FUNGIBLES`    | Renamed  |
| `CREATE_PROOF_FROM_AUTH_ZONE_OF_ALL`    | New  |

### CODE REVIEW

File: transaction/src/model/instruction.rs [hunk 1]
```diff
@@ -70,21 +70,36 @@ pub enum Instruction {
     },
 
     /// Creates a proof from the auth zone, by the given amount
-    CreateProofFromAuthZoneByAmount {
+    CreateProofFromAuthZoneOfAmount {
```
Comment (dhedey) [inline transaction/src/model/instruction.rs:73]: Would you consider putting resource address first in these instructions for consistency with transfer etc (as per my comment in the other PR?)

Comment (dhedey) [reply transaction/src/model/instruction.rs:73]: (And eg consistency with account's `create_proof_of_amount`)

File: radix-engine/src/blueprints/resource/auth_zone.rs [hunk 2]
```diff
@@ -206,26 +151,59 @@ impl AuthZoneBlueprint {
...
+        let composed_proof = compose_proof_by_amount(&proofs, resource_address, None, api)?;
+
+        let blueprint_name = match &composed_proof {
+            ComposedProof::Fungible(..) => FUNGIBLE_PROOF_BLUEPRINT,
+            ComposedProof::NonFungible(..) => NON_FUNGIBLE_PROOF_BLUEPRINT,
+        };
+        api.field_lock_write_typed(auth_zone_handle, &auth_zone)?;
+
+        let node_id = api.kernel_allocate_node_id(EntityType::InternalGenericComponent)?;
+        api.kernel_create_node(
```
Comment (dhedey) [inline radix-engine/src/blueprints/resource/auth_zone.rs:187]: Why can we use `new_simple_object` in `FungibleBucket`'s `create_proof_of_amount`, but not here? Is the former missing an outer_object? Or does `new_simple_object` just call a `new` constructor?

I also wonder if some of this logic could live / be commonized in the proof blueprint to reduce the risk of it diverging? :)

Comment (dhedey) [reply radix-engine/src/blueprints/resource/auth_zone.rs:187]: (appreciate this was just moved in this PR, so happy if we come back to this later)

File: radix-engine-tests/tests/auth_zone.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        manifest,
+        vec![NonFungibleGlobalId::from_public_key(&public_key)],
+    );
+
+    // Assert
+    receipt.expect_commit_success();
+}
+
+#[test]
+fn test_auth_zone_create_proof_by_all_for_non_fungible() {
```
Comment (dhedey) [inline radix-engine-tests/tests/auth_zone.rs:29]: `of_all`*?

File: radix-engine-tests/tests/auth_zone.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+use radix_engine::types::*;
+use scrypto_unit::*;
+use transaction::builder::ManifestBuilder;
+
+#[test]
+fn test_auth_zone_create_proof_by_all_for_fungible() {
```
Comment (dhedey) [inline radix-engine-tests/tests/auth_zone.rs:6]: `of_all`*?

File: radix-engine-tests/tests/blueprints/proof_creation/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+        pub fn create_fungible_vault() -> Vault {
+            Vault::with_bucket(Self::create_fungible_bucket())
+        }
+
+        pub fn create_non_fungible_vault() -> Vault {
+            Vault::with_bucket(Self::create_non_fungible_bucket())
+        }
+
+        pub fn prepare_fungible_proof() -> Bucket {
```
Comment (dhedey) [inline radix-engine-tests/tests/blueprints/proof_creation/src/lib.rs:262]: Possibly `prepare_auth_zone_fungible_proof_to_bucket` would be clearer, I expected this function to return a proof.

**Review by dhedey — APPROVED** [review]
Really nice work. Another very consistent / careful PR 👍 

Main question is about whether we should swap the parameter order too? I'd personally vote yes for consistency. Happy for it to be done in a separate PR though

File: transaction/src/model/instruction.rs (hunk 1 above)
Comment (iamyulong) [reply transaction/src/model/instruction.rs:73]: I will have no problem with this.

A much larger clean-up will be required to fix all these parameter order inconsistency.

File: radix-engine/src/blueprints/resource/auth_zone.rs (hunk 2 above)
Comment (iamyulong) [reply radix-engine/src/blueprints/resource/auth_zone.rs:187]: Let's revisit this as part of the "removing dependency on kernel api from native blueprints".

**Pull request merged**